packages/controller-utils/src/create-service-policy.test.ts

```diff
@@ -0,0 +1,2703 @@
+            const policy = createServicePolicy({
+              maxConsecutiveFailures,
+              onBreak,
+            });
+
+            const promise = policy.execute(() => mockService());
+            clock.runAllAsync().catch(() => {
+              // ignore any errors - adding to the promise queue is enough
+            });
+            await expect(promise).rejects.toThrow(
+              new Error(
+                'Execution prevented because the circuit breaker is open',
+              ),
+            );
+          });
+
+          it('calls the onBreak callback once with the error', async () => {
+            const maxConsecutiveFailures = DEFAULT_MAX_RETRIES;
+            let invocationCounter = 0;
+            const error = new Error('failure');
+            const mockService = () => {
+              invocationCounter += 1;
+              if (invocationCounter === DEFAULT_MAX_RETRIES + 1) {
+                return { some: 'data' };
+              }
+              throw error;
+            };
+            const onBreak = jest.fn();
+            const policy = createServicePolicy({
+              maxConsecutiveFailures,
+              onBreak,
+            });
+
+            const promise = policy.execute(() => mockService());
+            clock.runAllAsync().catch(() => {
+              // ignore any errors - adding to the promise queue is enough
+            });
+            try {
+              await promise;
+            } catch {
+              // ignore the error
+            }
+
+            expect(onBreak).toHaveBeenCalledTimes(1);
+            expect(onBreak).toHaveBeenCalledWith({ error });
+          });
+
+          it('does not call the onDegraded callback', async () => {
+            const maxConsecutiveFailures = DEFAULT_MAX_RETRIES;
+            let invocationCounter = 0;
+            const error = new Error('failure');
+            const mockService = () => {
+              invocationCounter += 1;
+              if (invocationCounter === DEFAULT_MAX_RETRIES + 1) {
+                return { some: 'data' };
+              }
+              throw error;
+            };
+            const onDegraded = jest.fn();
+            const policy = createServicePolicy({
+              maxConsecutiveFailures,
+              onDegraded,
+            });
+
+            const promise = policy.execute(() => mockService());
+            clock.runAllAsync().catch(() => {
+              // ignore any errors - adding to the promise queue is enough
+            });
+            try {
+              await promise;
+            } catch {
+              // ignore the error
+            }
+
+            expect(onDegraded).not.toHaveBeenCalled();
+          });
+
+          describe(`using the default circuit break duration (${DEFAULT_CIRCUIT_BREAK_DURATION})`, () => {
+            it('returns what the service returns if it is successfully called again after the circuit break duration has elapsed', async () => {
```

Comment:
This is a bit buried, but this (and other tests like it) is where we ensure that once the circuit enters a half-open state it can resolve to closed again after the circuit break duration has elapsed.